### PR TITLE
This one WEIRD trick will reduce new bot code by 90%

### PR DESCRIPTION
**Pull request by colin99d**
# Description

@DidierRLopes still hasn't made an AOL bot so I decided to take away his excuses. Thanks to some genius from @jose-donato we have abstracted the interface for bots. This means new bots just need a function for sending messages and sending images. So easy!!!

What has been/needs done:
- [x] Create a function to abstract new bot creation
- [x] Refactor telegram bot
- [x] Refactor groupme bot
- [x] Refactor slack bot
- [x] standardize help message
- [ ] have the logging code for each platform in the same file

# How has this been tested?

* Please describe the tests that you ran to verify your changes. 
* Provide instructions so we can reproduce. 
* Please also list any relevant details for your test configuration.

# Checklist:

- [x] Update [our Hugo documentation](https://gamestonkterminal.github.io/GamestonkTerminal/) following [these guidelines](https://github.com/GamestonkTerminal/GamestonkTerminal/tree/main/website).
- [x] Update our tests following [these guidelines](https://github.com/GamestonkTerminal/GamestonkTerminal/tree/main/tests).
- [x] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/GamestonkTerminal/GamestonkTerminal/blob/main/CONTRIBUTING.md).
- [x] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/GamestonkTerminal/GamestonkTerminal/tree/main/scripts).

# Others
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code passes all the checks pylint, flake8, black, ... To speed up development you should run `pre-commit install`.
- [x] New and existing unit tests pass locally with my changes. You can test this locally using `pytest tests/...`.

### CODE REVIEW

**Comment by jmaslek** [comment]
So can I do AIM yet

**Comment by DidierRLopes** [comment]
what have I done this time? :(

**Comment by colin99d** [comment]
This code has known bugs, but @tehcoderer has a PR that fixes most of them. Will do a run through once both PRs are merged to see which commands still have issues.

File: bots/common/helpers.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                        val = bool(val)
+                    if (isinstance(required, List) and val not in required) or (
+                        isinstance(required, Pattern) and not required.match(val)
+                    ):
+                        syntax = get_syntax(selected, full_cmd[0])
+                        sender(f"{syntax}\nInvalid argument for: {req_name}")
+                        get_arguments(selected, req_name, sender)
+                        return False
+                    other_args[req_name] = val
+                func = selected["function"]
```
Comment (tehcoderer) [inline bots/common/helpers.py:106]:
```py
                if (isinstance(required, List) and val not in required) or (
                    isinstance(required, Pattern) and not required.match(val)
                ):
                    syntax = get_syntax(selected, full_cmd[0])
                    bot.reply_to(message, f"{syntax}\nInvalid argument for: {req_name}")
                    get_arguments(selected, req_name, message)
                    return False
                elif isinstance(val, str) and req_name == "interval":
                    val = int(val)
                elif isinstance(val, str) and req_name == "strike":
                    val = float(val)
                other_args[req_name] = val
            func = selected["function"]
 ```

In my tests adding this fixed the int/float type issues    🚀

File: bots/common/commands_dict.py [hunk 2]
```diff
@@ -544,23 +543,23 @@ def get_tickers() -> List[str]:
         "required": {"ticker": tickers, "interval": opt_intervals},
         "optional": {"past_days": re_int, "start": re_date, "end": re_date},
     },
-    "opt-overview": {
+    "opt_overview": {
         "function": overview_opt_command,
         "required": {"ticker": tickers, "expiry": re_date},
```
Comment (tehcoderer) [inline bots/common/commands_dict.py:548]: I was able to get the expiry dates args to work only after I changed `re_date` to `re_name`. Issue with the re.Pattern?

**Review by tehcoderer — APPROVED** [review]
Great work with the helpers! Can't wait for some samsung fridge bots !

File: bots/common/helpers.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+def send_options(name: str, items: Union[List[Any], Set[Any]], sender: Any) -> None:
+    """Sends the options for a user
+
+    Parameters
+    ----------
+    name : str
+        The name of the section
+    items : List[str]
+        The items the user can select from
+    message: Any
```
Comment (jose-donato) [inline bots/common/helpers.py:17]: wrong sender docstring

**Review by jose-donato — CHANGES_REQUESTED** [review]
`bots/ta_kc.png` i assume this was also added by mistake

File: bots/common/helpers.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            selections = f"{selections[:990]}"
+        sender(f"Options: {selections}")
+
+
+def non_slash(text, sender, showview):
+    cmd = text[1:]
+    full_cmd = cmd.split("/")
+    group = full_cmd[0].split("_")[0]
+    parents = {x.split("_")[0] for x in available_commands}
+    if text[0] == "!":
```
[the remark this answers is not in the document]
Comment (colin99d) [reply bots/common/helpers.py:92]: Bet.

[the remark this answers is not in the document]
Comment (DidierRLopes) [reply bots/common/helpers.py]: > ```python
> ```python
> if text[0] == "!" or "/":
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> Would this work? haha
> ```

In order to achieve this you would want to do:
`if text[0] in ("!", "/"):`

**Review by jose-donato — APPROVED** [review]

**Pull request merged**